src/type.js

```diff
@@ -167,106 +158,215 @@ async function typeImpl(
       }
     }
   }
+}
 
-  function fireInputEventIfNeeded({
-    newValue,
-    newSelectionStart,
-    eventOverrides,
-  }) {
-    const prevValue = currentValue()
-    if (!currentElement().readOnly && newValue !== prevValue) {
-      fireEvent.input(currentElement(), {
-        target: {value: newValue},
-        ...eventOverrides,
-      })
+function getNextCallback(remainingString, skipAutoClose) {
+  const modifierCallback = getModifierCallback(remainingString, skipAutoClose)
+  if (modifierCallback) {
+    return modifierCallback
+  }
 
-      setSelectionRange({newValue, newSelectionStart})
-    }
+  const specialCharCallback = getSpecialCharCallback(remainingString)
+  if (specialCharCallback) {
+    return specialCharCallback
+  }
+
+  return getTypeCallback(remainingString)
+}
 
-    return {prevValue}
+function getModifierCallback(remainingString, skipAutoClose) {
+  const modifierKey = Object.keys(modifierCallbackMap).find(key =>
+    remainingString.startsWith(key),
+  )
+  if (!modifierKey) {
+    return null
   }
+  const callback = modifierCallbackMap[modifierKey]
 
-  function typeCharacter(
-    char,
-    {
-      prevWasMinus = false,
-      prevWasPeriod = false,
-      prevValue = '',
-      eventOverrides,
-    },
+  // if this modifier has an associated "close" callback and the developer
+  // doesn't close it themselves, then we close it for them automatically
+  // Effectively if they send in: '{alt}a' then we type: '{alt}a{/alt}'
+  if (
+    !skipAutoClose &&
+    callback.closeName &&
+    !remainingString.includes(callback.closeName)
   ) {
-    const key = char // TODO: check if this also valid for characters with diacritic markers e.g. úé etc
-    const keyCode = char.charCodeAt(0)
-    let nextPrevWasMinus, nextPrevWasPeriod
+    remainingString += callback.closeName
+  }
+  remainingString = remainingString.slice(modifierKey.length)
+  return {
+    callback,
+    remainingString,
+  }
+}
 
-    const keyDownDefaultNotPrevented = fireEvent.keyDown(currentElement(), {
-      key,
-      keyCode,
-      which: keyCode,
+function getSpecialCharCallback(remainingString) {
+  const specialChar = Object.keys(specialCharCallbackMap).find(key =>
+    remainingString.startsWith(key),
+  )
+  if (!specialChar) {
+    return null
+  }
+  return {
+    callback: specialCharCallbackMap[specialChar],
+    remainingString: remainingString.slice(specialChar.length),
+  }
+}
+
+function getTypeCallback(remainingString) {
+  const character = remainingString[0]
+  const callback = createTypeCharacter(character)
```

Comment:
I'm not sure we need the createTypeCharacter function. This is a quick and easy one-liner.

```suggestion
  const callback = context => typeCharacter(character, context)
```

Reply:
Sorry, just had time to read this. That's true. Initially I used createTypeCharacter in more places, but now it is a bit too much 😄